File: store/src/main/java/com/nytimes/android/external/store3/base/impl/CacheFactory.java
package com.nytimes.android.external.store3.base.impl;

import com.nytimes.android.external.cache3.Cache;
import com.nytimes.android.external.cache3.CacheBuilder;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;

public final class CacheFactory {
    private CacheFactory() {

    }

    static <Key, Parsed> Cache<Key, Parsed> createCache(MemoryPolicy memoryPolicy) {
        return createBaseCache(memoryPolicy);
    }

    static <Key, Parsed> Cache<Key, Parsed> createInflighter(MemoryPolicy memoryPolicy) {
        return createBaseInFlighter(memoryPolicy);
    }

    public static <Key, Parsed> Cache<Key, Observable<Parsed>> createRoomCache(MemoryPolicy memoryPolicy) {
        return createBaseCache(memoryPolicy);
    }


    public static <Key, Parsed> Cache<Key, Observable<Parsed>> createRoomInflighter(MemoryPolicy memoryPolicy) {
        return createBaseInFlighter(memoryPolicy);
    }


    private static <Key, Value> Cache<Key, Value> createBaseInFlighter(MemoryPolicy memoryPolicy) {
        long expireAfterToSeconds = memoryPolicy == null ? StoreDefaults.getCacheTTLTimeUnit()
                .toSeconds(StoreDefaults.getCacheTTL())
                : memoryPolicy.getExpireAfterTimeUnit().toSeconds(memoryPolicy.getExpireAfterWrite());
        long maximumInFlightRequestsDuration = TimeUnit.MINUTES.toSeconds(1);

        if (expireAfterToSeconds > maximumInFlightRequestsDuration) {
            return CacheBuilder
                    .newBuilder()
                    .expireAfterWrite(maximumInFlightRequestsDuration, TimeUnit.SECONDS)
                    .build();
        } else {
            long expireAfter = memoryPolicy == null ? StoreDefaults.getCacheTTL() :
                    memoryPolicy.getExpireAfterWrite();
            TimeUnit expireAfterUnit = memoryPolicy == null ? StoreDefaults.getCacheTTLTimeUnit() :
                    memoryPolicy.getExpireAfterTimeUnit();
            return CacheBuilder.newBuilder()
                    .expireAfterWrite(expireAfter, expireAfterUnit)
                    .build();
        }
    }


    private static <Key, Value> Cache<Key, Value> createBaseCache(MemoryPolicy memoryPolicy) {
        if (memoryPolicy == null) {
            return CacheBuilder
                    .newBuilder()
                    .maximumSize(StoreDefaults.getCacheSize())
                    .expireAfterWrite(StoreDefaults.getCacheTTL(), StoreDefaults.getCacheTTLTimeUnit())
                    .build();
        } else {
            if (memoryPolicy.getExpireAfterAccess() == memoryPolicy.DEFAULT_POLICY) {
                return CacheBuilder
                        .newBuilder()
                        .maximumSize(memoryPolicy.getMaxSize())
                        .expireAfterWrite(memoryPolicy.getExpireAfterWrite(), memoryPolicy.getExpireAfterTimeUnit())
                        .build();
            } else {
                return CacheBuilder
                        .newBuilder()
                        .maximumSize(memoryPolicy.getMaxSize())
                        .expireAfterAccess(memoryPolicy.getExpireAfterAccess(), memoryPolicy.getExpireAfterTimeUnit())
                        .build();
            }
        }
    }

}

File: store/src/main/java/com/nytimes/android/external/store3/base/impl/CacheFactory.kt
package com.nytimes.android.external.store3.base.impl

import com.nytimes.android.external.cache3.Cache
import com.nytimes.android.external.cache3.CacheBuilder
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

object CacheFactory {

    internal fun <Key, Parsed> createCache(memoryPolicy: MemoryPolicy?): Cache<Key, Parsed> {
        return createBaseCache(memoryPolicy)
    }

    internal fun <Key, Parsed> createInflighter(memoryPolicy: MemoryPolicy?): Cache<Key, Parsed> {
        return createBaseInFlighter(memoryPolicy)
    }

    @JvmStatic
    fun <Key, Parsed> createRoomCache(memoryPolicy: MemoryPolicy): Cache<Key, Observable<Parsed>> {
        return createBaseCache(memoryPolicy)
    }

    @JvmStatic
    fun <Key, Parsed> createRoomInflighter(memoryPolicy: MemoryPolicy): Cache<Key, Observable<Parsed>> {
        return createBaseInFlighter(memoryPolicy)
    }

    private fun <Key, Value> createBaseInFlighter(memoryPolicy: MemoryPolicy?): Cache<Key, Value> {
        val expireAfterToSeconds = memoryPolicy?.expireAfterTimeUnit?.toSeconds(memoryPolicy.expireAfterWrite)
                ?: StoreDefaults.getCacheTTLTimeUnit()
                        .toSeconds(StoreDefaults.getCacheTTL())
        val maximumInFlightRequestsDuration = TimeUnit.MINUTES.toSeconds(1)

        return if (expireAfterToSeconds > maximumInFlightRequestsDuration) {
            CacheBuilder
                    .newBuilder()
                    .expireAfterWrite(maximumInFlightRequestsDuration, TimeUnit.SECONDS)
                    .build()
        } else {
            val expireAfter = memoryPolicy?.expireAfterWrite ?: StoreDefaults.getCacheTTL()
            val expireAfterUnit = if (memoryPolicy == null)
                StoreDefaults.getCacheTTLTimeUnit()
            else
                memoryPolicy.expireAfterTimeUnit
            CacheBuilder.newBuilder()
                    .expireAfterWrite(expireAfter, expireAfterUnit)
                    .build()
        }
    }


    private fun <Key, Value> createBaseCache(memoryPolicy: MemoryPolicy?): Cache<Key, Value> {
        return if (memoryPolicy == null) {
            CacheBuilder
                    .newBuilder()
                    .maximumSize(StoreDefaults.getCacheSize())
                    .expireAfterWrite(StoreDefaults.getCacheTTL(), StoreDefaults.getCacheTTLTimeUnit())
                    .build()
        } else {
            if (memoryPolicy.expireAfterAccess == MemoryPolicy.DEFAULT_POLICY) {
                CacheBuilder
                        .newBuilder()
                        .maximumSize(memoryPolicy.maxSize)
                        .expireAfterWrite(memoryPolicy.expireAfterWrite, memoryPolicy.expireAfterTimeUnit)
                        .build()
            } else {
                CacheBuilder
                        .newBuilder()
                        .maximumSize(memoryPolicy.maxSize)
                        .expireAfterAccess(memoryPolicy.expireAfterAccess, memoryPolicy.expireAfterTimeUnit)
                        .build()
            }
        }
    }

}
